presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -495,18 +503,71 @@ private void invokeNoMoreSplitsIfNecessary()
+
+                // list all files in the partition
+                List<LocatedFileStatus> files = new ArrayList<>(partitionBucketCount);
+                try {
+                    Iterators.addAll(files, new HiveFileIterator(table, readPath, fs, directoryLister, namenodeStats, FAIL, ignoreAbsentPartitions));
+                }
+                catch (HiveFileIterator.NestedDirectoryNotAllowedException e) {
+                    // Fail here to be on the safe side. This seems to be the same as what Hive does
+                    throw new PrestoException(
+                            HIVE_INVALID_BUCKET_FILES,
+                            format("Hive table '%s' is corrupt. Found sub-directory in bucket directory for partition: %s",
+                                    table.getSchemaTableName(),
+                                    splitFactory.getPartitionName()));
+                }
+                lastResult = hiveSplitSource.addToQueue(getBucketedSplits(files, splitFactory, tableBucketInfo.get(), bucketConversion, splittable, acidInfoBuilder.build()));
             }
+
+            for (HadoopShims.HdfsFileStatusWithId hdfsFileStatusWithId : fileStatusOriginalFiles) {
+                List<LocatedFileStatus> locatedFileStatuses = new ArrayList<>();
+                locatedFileStatuses.add((LocatedFileStatus) hdfsFileStatusWithId.getFileStatus());
+                lastResult = hiveSplitSource.addToQueue(getBucketedSplits(locatedFileStatuses, splitFactory, tableBucketInfo.get(), bucketConversion, splittable,
+                        acidInfoBuilder
+                        .buildWithRequiredOriginalFiles(getBucketNumber(hdfsFileStatusWithId.getFileStatus().getPath().getName()).getAsInt())));
+            }
+
             return lastResult;
         }
 
         for (Path readPath : readPaths) {
-            fileIterators.addLast(createInternalHiveSplitIterator(readPath, fs, splitFactory, splittable, acidInfo));
+            fileIterators.addLast(createInternalHiveSplitIterator(readPath, fs, splitFactory, splittable, acidInfoBuilder.build()));
+        }
+
+        if (!fileStatusOriginalFiles.isEmpty()) {
+            generateOriginalFilesSplits(fs, splitFactory, fileStatusOriginalFiles, acidInfoBuilder);
         }
 
         return COMPLETED_FUTURE;
     }
 
+    private void generateOriginalFilesSplits(
+            FileSystem fs,
+            InternalHiveSplitFactory splitFactory,
+            List<HadoopShims.HdfsFileStatusWithId> originalFileLocations,
+            AcidInfo.Builder acidInfoBuilder)
+    {
+        fileIterators.addLast(
+                originalFileLocations.stream()
+                        .map(file -> file.getFileStatus())
+                        .map(fileStatus -> {
+                            try {
+                                return splitFactory.createInternalHiveSplit(fileStatus,
```

Comment:
nit, per code style:

```suggestion
                                return splitFactory.createInternalHiveSplit(
                                    fileStatus,
```

Reply:
fixed.